### PR TITLE
minor fix for nonsquare, sparse, inplace jacobians

### PR DESCRIPTION
**Pull request by andrewning**
Line 373/412 fails for nonsquare matrices (length of fx1 and x1 are different).

### CODE REVIEW

File: src/jacobians.jl [hunk 1]
```diff
@@ -370,8 +370,7 @@ function finite_difference_jacobian!(
                 # Now return x1 back to its original value
                 ArrayInterface.allowed_setindex!(x1, x1_save, color_i)
             else # Perturb along the colorvec vector
-                @. fx1 = x1 * (_color == color_i)
-                tmp = norm(fx1)
+                tmp = norm(x1 .* (_color .== color_i))
```
Comment (YingboMa) [inline src/jacobians.jl:373]: This makes it allocating. Could you create a cache vector that has length `max(fx1, x1)`?

Comment (andrewning) [reply src/jacobians.jl:373]: That's a good point.  I don't think fx1 matters at all for this particular calculation, it just needs to be length(x1). However,  I'm not sure where you would want to cache it.  Could modify x1 in place, then recopy over the entries in x back into x1.  Or I suppose you could create another vector in JacobianCache.

Comment (andrewning) [reply src/jacobians.jl:373]: actually couldn't we do something simpler?  Seem like all we need is the one entry in x1 that corresponds to the current color.  Doesn't seem necessary to do the multiplication.  `tmp = abs(x1[_color .== color_i])`  or perhaps I'm missing another use case.

Comment (andrewning) [reply src/jacobians.jl]: scratch that, it still allocates.   The version below doesn't allocate.

**Comment by ChrisRackauckas** [comment]
This had to be reverted in https://github.com/JuliaDiff/FiniteDiff.jl/pull/129. Downstream we saw failures from inaccurate gradients. The step size was pretty far off. A downstream test was added to catch this. Feel free to try this again and see if it can keep the regression tests passing with a few tweaks.

**Pull request merged**